File: 链表/list_test.java
import java.util.*;

public class list_test {
    public static void main(String[] args) {
        List l=new List();
        l.add(1);
        l.add(2);
        l.printList();

        l.insert(3, 1);
        l.printList();

        l.remove(4);
        l.printList();
    }
}

class List {
    private Node firstNode;
    
    public List(){
        init();
    }

    public void init() {
        firstNode=new Node(null);
    }

    //判断链表是否为空
    public boolean isEmpty(){
        if(firstNode==null) 
            return true;
        else 
            return false;
    }


    //在链表末尾插入新节点
    public void add(Object data) {
        Node node=new Node(data);

        Node pointer = firstNode;
        while (pointer.next() != null) {
            pointer = pointer.next();
        }
        pointer.setNext(node);

    }

    //插入节点
    public void insert(Object data,int i){
        Node node=new Node(data);            
        Node pointer=firstNode;
        int position=0;
        while (position<i && pointer!=null) {
            pointer=pointer.next();
            ++position;
        }

        if(pointer==null||i<0){
            System.out.println("invalid position");
            return;
        }
        node.setNext(pointer.next());
        pointer.setNext(node);
    }

    //删除节点
    public void remove(int position){
        Node pointer=firstNode;
        int i=1;
        while (i<position && pointer!=null) {
            pointer=pointer.next();
            ++i;
        }

        if(pointer==null||position<1||pointer.next()==null){
            System.out.println("invalid position");
            return;
        }

        pointer.setNext(pointer.next().next());
    }

    public void printList(){
        Node pointer=this.firstNode.next();
        while (pointer!=null) {
            System.out.print(pointer.getData()+"  ");
            pointer=pointer.next();
        }
        System.out.println();
    }

}

/**
 * Node
 */
class Node {

    private Object data;
    private Node next;

    public Node(Object data){
        this.data=data;
        this.next=null;
    }

    public Node(Object data,Node next){
        this.data=data;
        this.next=next;
    }

    public Node next(){
        return this.next;
    }

    public void setNext(Node next){
        this.next=next;
    }

    public Object getData(){
        return this.data;
    }

    public void setData(Object data){
        this.data=data;
    }

}

File: 链表/list_test.c
#include <stdio.h>
#include <stdlib.h>
typedef int Datatype;
typedef struct Node
{
    Datatype data;
    struct Node *next;
}Node,*first;

Node *initlist();
void add(Node *list, Datatype data);
void printList(Node *list);
void insert(Node *list,Datatype data, int position);
void removeAt(Node *list,int position);
void destroy(Node *list);
void listDestroy(Node *head);

int main(int argc, char const *argv[])
{
    Node *list=initlist();
    add(list,1);
    add(list,2);
    insert(list,3,2);
    printList(list);
    removeAt(list,2);
    printList(list);printf("%d",list==NULL);
    destroy(list);
    printf("%d",list==NULL);
    //listDestroy(list);
    //printList(list);
    return 0;
}

Node *initlist(){
    Node *first=(Node *)malloc(sizeof(Node));
    first->next=NULL;
    return first;
}

void add(Node *list, Datatype data){
    Node *pointer=list;
    Node *node=(Node *)malloc(sizeof(Node));
    node->data=data;
    
    while(pointer->next!=NULL){
        pointer=pointer->next;
    }

    pointer->next=node;
    node->next=NULL;
}

void printList(Node *list){
    if(list==NULL){
        printf("Empty list!");
        return;
    }

    Node *pointer=list->next;
    
    while(pointer!=NULL){
        printf("%d  ",pointer->data);
        pointer=pointer->next;
    }
    printf("\n");
}

void insert(Node *list,Datatype data, int position){
    Node *pointer=list;
    int i=0;
    while(i<position && pointer!=NULL){
        pointer=pointer->next;
        ++i;
    }

    if(position<0||pointer==NULL){
        return;
    }

    Node *node=(Node *)malloc(sizeof(Node));
    node->data=data;
    node->next=pointer->next;
    pointer->next=node;
}

void removeAt(Node *list,int position){
    Node *pointer=list;
    int i=1;
    while(i<position && pointer!=NULL){
        pointer=pointer->next;
        ++i;
    }

    if(position<1||pointer==NULL||pointer->next==NULL)
        return;

    Node *node=pointer->next;
    pointer->next=node->next;
    free(node);
}

//销毁列表
void destroy(Node *list){
    Node *p=list;    
    while(list!=NULL){
        p=list;
        list=list->next;
        free(p);  
        printf("______\n");   
    }
    printf("%d",list==NULL);
}
